### ISSUE TITLE
[Refactor] use str_starts_with, str_ends_with

### ISSUE THREAD

**Issue by fiwswe** [dokuwiki/dokuwiki#4060]
Many places in DokuWiki use patterns designed to check the start or end of a string against a value. Currently [substr](https://www.php.net/manual/en/function.substr) and [strncmp](https://www.php.net/manual/en/function.strncmp) seem to be used for this.

The PHP 8+ built-ins [str_starts_with](https://www.php.net/manual/en/function.str-starts-with) and [str_ends_with](https://www.php.net/manual/en/function.str-ends-with) make the code more readable because the intention of the code becomes clear immediately. And they avoid coding mistakes in these cases.

Thus I propose to change as many places in DokuWiki as possible to use the new functions.

To allow older versions of PHP to use these functions, polyfills where added in https://github.com/dokuwiki/dokuwiki/commit/a4c648fffdf5bc5db83c9aa9c6c4bf5fb5edc85f.

I am working on a PR (TBD) to implement this…

## Caveats

The polyfills are implemented in `./inc/compatibility.php`. Thus they are useable only after this file has been included/loaded. Currently this happens in https://github.com/dokuwiki/dokuwiki/blob/fa4b9dc1e9fc6a931deee2786437a3d4f9a8ef71/inc/load.php#L39 which is rather late IMHO.

`./inc/load.php` is included/loaded in https://github.com/dokuwiki/dokuwiki/blob/fa4b9dc1e9fc6a931deee2786437a3d4f9a8ef71/inc/init.php#L207 again rather late.

And `./inc/init.php` is included from many files, i.e. those scripts that are entry points for DokuWiki:
* https://github.com/dokuwiki/dokuwiki/blob/fa4b9dc1e9fc6a931deee2786437a3d4f9a8ef71/doku.php#L38 — loaded pretty early so good.
* https://github.com/dokuwiki/dokuwiki/blob/fa4b9dc1e9fc6a931deee2786437a3d4f9a8ef71/feed.php#L20 — loaded pretty early so good.
* https://github.com/dokuwiki/dokuwiki/blob/fa4b9dc1e9fc6a931deee2786437a3d4f9a8ef71/index.php#L20 — loaded pretty early so good.
* https://github.com/dokuwiki/dokuwiki/blob/fa4b9dc1e9fc6a931deee2786437a3d4f9a8ef71/install.php#L20 — loaded pretty early so good.
* `./bin/*.php` — loaded pretty early in all cases so good.
* `./lib/exe/*.php` — loaded pretty early in all cases so good.
* https://github.com/dokuwiki/dokuwiki/blob/fa4b9dc1e9fc6a931deee2786437a3d4f9a8ef71/lib/plugins/styling/popup.php#L4 — loaded pretty early so good.
* https://github.com/dokuwiki/dokuwiki/blob/fa4b9dc1e9fc6a931deee2786437a3d4f9a8ef71/lib/tpl/index.php#L16 — loaded pretty early so good.

In summary we need to check any code that might be executed (not evaluated) before the polyfills are loaded and avoid using the polyfills there. Otherwise all of DokuWiki can use the new functions. It might be prudent to move loading `./inc/compatibility.php` to an earlier point to avoid issues. Thoughts?

## Other considerations

3rd-party plugins and templates, i.e. additions not present in the base install of DokuWiki, can and should also switch to use the new functions. However they may need to provide their own polyfills to stay compatible with DokuWiki ≤2023-04-04a. These changes need to be addressed in the repositories of the plugins/templates in question and are out-of-scope here.

**Comment by fiwswe** [comment]
Where can the new functions **_not_** be used?
(Note: All of these issues go away with PHP versions ≥8. But as long as PHP 7.4 is still supported in DokuWiki we need to deal with this.)
* `./preload.php`
  * Is loaded very early, long before `./inc/compatibility.php` is loaded.
  * But this file is optional and user-supplied. So this is not much of an issue.
* Any `./config/*.php` files
  * Test are loaded very early, long before `./inc/compatibility.php` is loaded.
  * Not an issue as these files tend to contain data (PHP arrays) not executable code.
* `./inc/config_cascade.php`
  * Is loaded very early, long before `./inc/compatibility.php` is loaded.
  * Contains data (PHP array) only.
* `./inc/init.php`
  * `./inc/compatibility.php` is loaded in `./inc/load.php` which is loaded from here. Thus code that might execute before that has happened can not use the new functions.
  * This file does contain several candidates for replacement. Unless the loading of `./inc/compatibility.php` is moved up to around line 28 of this file, these can't be replaced. And at that point `DOKU_INC` has not been defined yet (though we could ignore that and use `require_once 'compatibility.php';` as the file is in the same directory as `./inc/init.php`).
    * Loading `.inc/compatibility.php` early here would also allow the use of the polyfills in the files mentioned above. I'll probably leave that for a future PR for now, unless I get endorsements to do this now?

**Comment by Klap-in** [comment]
For adoption in plugins/templates I think it is for their authors in general easier to wait a release longer, our [depreciation](https://www.dokuwiki.org/devel:releases#deprecation) policy is two releases, and by waiting a release you do not need all that fallback code. (I guess that deployment of php8 and higher become more common as well, which could in this case help adoption of these functions a bit.)

If we start adding more replacements to rector, it might be useful to be able for a plugin to apply Dokuwiki specific rector rules per Dokuwiki-release. Encouraging always directly applying new stuff or depreciations, means the backward compatibility of plugins becomes bad, or we/authors have to introduce a lot of temporary fallback code, which an author could remove typically a release later. Therefore, I’m in favor (in general) to wait one release before encouraging(“enforcing”) cleanup due to depreciations/new additions.

**Comment by Klap-in** [comment]
> The polyfills are implemented in `./inc/compatibility.php`. Thus they are useable only after this file has been included/loaded. Currently this happens in …

My first view is that especially the functions in init.php should not use these polyfills yet.

**Comment by fiwswe** [comment]
> My first view is that especially the functions in init.php should not use these polyfills yet.

Ok, that is fine. Such a drastic change in load order might have unintended consequences in the future. So I'll leave that alone for now.

**Comment by fiwswe** [comment]
Ok, `./inc/*` is done. I'll look at the rest (excepting `./lib/vendor/*`) next.

PR #4062
